File: packages/orchestrator/src/toon/compression.ts
export const COMPRESSION_THRESHOLD = 1024;
const MAGIC_COMPRESSED = 0x5a;

export function isCompressed(buffer: Uint8Array): boolean {
  return buffer.length > 0 && buffer[0] === MAGIC_COMPRESSED;
}

function simpleRLEEncode(data: Uint8Array): Uint8Array {
  const result: number[] = [];
  let i = 0;
  while (i < data.length) {
    const byte = data[i] ?? 0;
    let count = 1;
    while (i + count < data.length && data[i + count] === byte && count < 255) {
      count++;
    }
    if (count > 3 || byte === 0xff) {
      result.push(0xff, byte, count);
    } else {
      for (let j = 0; j < count; j++) result.push(byte);
    }
    i += count;
  }
  return new Uint8Array(result);
}

function simpleRLEDecode(data: Uint8Array): Uint8Array {
  const result: number[] = [];
  let i = 0;
  while (i < data.length) {
    if (data[i] === 0xff && i + 2 < data.length) {
      const byte = data[i + 1] ?? 0;
      const count = data[i + 2] ?? 0;
      for (let j = 0; j < count; j++) result.push(byte);
      i += 3;
    } else {
      result.push(data[i] ?? 0);
      i++;
    }
  }
  return new Uint8Array(result);
}

export function compress(data: Uint8Array): Uint8Array {
  if (data.length < COMPRESSION_THRESHOLD) return data;
  const encoded = simpleRLEEncode(data);
  if (encoded.length >= data.length) return data;
  const result = new Uint8Array(1 + encoded.length);
  result[0] = MAGIC_COMPRESSED;
  result.set(encoded, 1);
  return result;
}

export function decompress(data: Uint8Array): Uint8Array {
  if (!isCompressed(data)) return data;
  return simpleRLEDecode(data.subarray(1));
}

File: packages/ai-engine/src/toon/compression.py
"""
TOON RLE compression — Python port of packages/orchestrator/src/toon/compression.ts

Uses a simple run-length encoding scheme with a 0xFF escape byte.
Runs of >3 identical bytes (or any run of 0xFF) are encoded as [0xFF, byte, count].
A single magic byte (0x5A) prefix marks compressed payloads.
"""

COMPRESSION_THRESHOLD = 1024
MAGIC_COMPRESSED = 0x5A


def is_compressed(data: bytes) -> bool:
    return len(data) > 0 and data[0] == MAGIC_COMPRESSED


def _simple_rle_encode(data: bytes) -> bytes:
    result = bytearray()
    i = 0
    length = len(data)

    while i < length:
        byte = data[i]
        count = 1
        while i + count < length and data[i + count] == byte and count < 255:
            count += 1

        if count > 3 or byte == 0xFF:
            result.append(0xFF)
            result.append(byte)
            result.append(count)
        else:
            for _ in range(count):
                result.append(byte)

        i += count

    return bytes(result)


def _simple_rle_decode(data: bytes) -> bytes:
    result = bytearray()
    i = 0
    length = len(data)

    while i < length:
        if data[i] == 0xFF and i + 2 < length:
            byte = data[i + 1]
            count = data[i + 2]
            for _ in range(count):
                result.append(byte)
            i += 3
        else:
            result.append(data[i])
            i += 1

    return bytes(result)


def compress(data: bytes) -> bytes:
    if len(data) < COMPRESSION_THRESHOLD:
        return data
    encoded = _simple_rle_encode(data)
    if len(encoded) >= len(data):
        return data
    return bytes([MAGIC_COMPRESSED]) + encoded


def decompress(data: bytes) -> bytes:
    if not is_compressed(data):
        return data
    return _simple_rle_decode(data[1:])
